Chore10 schema validation

Kirosoft opened the pull request:
New internal schema validation tool to:-

1) check the performance of reg ex expressions within our json schema
2) check the performance of the overall schema against supplied data

The tool return 0 on success or non-zero for failure (so it can be used as part of the automated tests via jenkins etc)

Kirosoft committed:
Initial commit for schema validation tool

Kirosoft committed:
flake8 and improve output formatting

Kirosoft committed:
refactor some function names for readability

Kirosoft committed:
added some extra payload tests

Kirosoft committed:
added execution script for use via jenkins
refactored some generally useful functions into /application/deed/utils

endamccormack commented on schema_tests.py:
```diff
@@ -0,0 +1,71 @@
+
+
+def verify_pattern(element, key, obj, idx):
+    print("Checking patterns for '%s'" % key)
```
do you think we should use the new logging mechanism to replace prints?

e.g.

```
import logging
from logger import logging_config

logging_config.setup_logging()
LOGGER = logging.getLogger(__name__)

LOGGER.info("Starting the server")
```

Kirosoft replied:
Is a good thought. I am not sure since this is part of the tool execution logs. Do the unit test currently output to the same system logs?

tsangsimon replied:
I agree it would be better to use the new logging mechanism as that writes both to the console and to logs.

endamccormack replied:
I have no strong feelings either way on this one, the logger both outputs into a file so in a kibana readable way so we could build a dashboard on our unittest passes over time and it outputs to the console so developers can see whats going on. I have no strong feelings either way other than consistency and history. If you think its a good idea, then it might be worth doing if not then no. Your decision.

endamccormack commented on schema_tests/utils.py:
```diff
@@ -0,0 +1,17 @@
+import re
+
+
+def load_json_file(file_name):
```
file_name or file_path?

Kirosoft replied:
yes, should be file_path. thx

endamccormack commented:
Pretty excellent work. One question though, seems like the schema tests are really just unit tests and shouldn't really have their own directories, I think maybe they should be refactored to just run with the unit tests, I don't see much point in keeping them separate. What do you think?

Kirosoft commented:
I kept them separate because you might update the unit tests and not affect the schema tests and vice versa. Could be persuaded either way though. 
There are a couple of ways this could be done:

1) just add in the command into unit_test.sh
2) refactor to integrate as a pytest module (more work but probably tidier)

Not sure?

Kirosoft committed:
Moved schema_tests to be part of unit_tests
Moved some functions into application/deed/utils (they will be useful for the next task)
Removed the now redundant schema_tests/utils
Added unit test to run schema_tests

Kirosoft committed:
flake8

endamccormack commented:
Well thats just why I thought it should be just part of the unit test suite. In my mind it should run when the unit tests runs and it is only testing a unit really.  I agree with keeping them separate, but by that I think there should be a schema folder in the unit-test directory, not at the top level directory.

I think it makes more sense to a developer making a change. To me it just is a unit test. Schema test failures should be recorded at the same time that unit test failures are recorded, I would probably say in the same file. Its just testing a small piece of functionality.

Unit tests themselves are probably going to start breaking out the testing of each component into their own directories like we were doing in ALPHA to make it cleaner as that test base is going to be getting larger.

https://github.com/LandRegistry/charges-deed-api/tree/master/tests

Still of course open to better suggestions if you disagree.

nigelpain commented on unit_tests/schema_tests.json:
```diff
@@ -0,0 +1,579 @@
+          "payload": "",
+          "expected": false,
+          "description": "blank is not ok"
+        }
```
Just looking at the acceptance criteria for the old story and it's saying gender optional so having it as blank would be ok?

Kirosoft replied:
Currently, it is optional in the sense that you do not need to provide it. The schema just enforces that if you do provide it then it is valid?

nigelpain commented on unit_tests/schema_tests.json:
```diff
@@ -0,0 +1,579 @@
+          "payload": "0723 424340",
+          "expected": false,
+          "description": "starts with 07 but has a space"
+        }
```
Have we got a duplicate test here as we are checking that the validation fails in both cases due to the space present.

Kirosoft replied:
The second test is one digit shorter. I was just checking that the test was not just checking length and allowing a substitute space in the case where there were fewer digits. Description attached to the test could be clearer though.

Kirosoft committed:
flake8

nigelpain commented on unit_tests/schema_tests.json:
```diff
@@ -0,0 +1,579 @@
+          "payload": "PL21 0KK",
+          "expected": false,
+          "description": "invalid end characters"
+        },
```
Any need for this scenario as it seems to be just a duplicate of the one above using postcode PL21 0CC.

nigelpain commented on unit_tests/schema_tests.json:
```diff
@@ -0,0 +1,579 @@
+          "payload": "e-MD12 45",
+          "expected": false,
+          "description": "cannot have spaces"
+        }
```
Wondering that it might be worth while introducing a scenario to check the rule that only one alpha character can be on the end if three of four number e-MD reference, so feeding in something like e-MD1234AB and checking that fails validation.

Kirosoft replied:
ok

nigelpain commented on unit_tests/schema_tests.json:
```diff
@@ -0,0 +1,579 @@
+          "payload": "DN02222",
+          "expected": false,
+          "description": "must start with a 1-9 after initial chars"
+        },
```
This seems like a duplicate of the one directly above it unless I'm completely missing something here.

Kirosoft replied:
yep, looks like a duplicate :-) .. will remove

nigelpain commented on unit_tests/schema_tests.json:
```diff
@@ -0,0 +1,579 @@
+          "payload": "DN1234T",
+          "expected": false,
+          "description": "can end with t only when no starting alphas"
+        }
```
I would also look at adding a scenario that covers title numbers that end with a character that isn't Z or T just to verify that is being restricted.
Also would think about a scenario to test invalid number of alpha characters at the start of the title number such as ABCD1234.
Also also would look at validation the length of numbered section in the title number, so using a title number such as AB12345678 for example where there is more than five numbers.

Kirosoft replied:
ok will do

nigelpain commented:
Had a look through from a test coverage point of view and left a few comments, looks really good though, was expecting it to be less readable than it is, so that's really cool.

Phemus commented on unit_tests/schema_tests.json:
```diff
@@ -0,0 +1,579 @@
+            {
+                "forename": "Amanda",
+                "middle_name": "Elizabeth",
+                "surname": "Smy the",
```
Just thinking but should a space in a surname be a invalid scenario? as its perfectly possible to have a double barrel surname for various reasons

Kirosoft replied:
I would think double barrel names are hyphenated e.g. smith-jones

Phemus replied:
I'm not 100% sure on this but i have known people with non hyphenated double barrel names might be worth confirming the business rule on this

Phemus commented on unit_tests/schema_tests.json:
```diff
@@ -0,0 +1,579 @@
+                "phone_number": "07507154078"
+            }
+
+         ]
```
is this indented correctly? or should the array level be one tab further in?

Kirosoft replied:
Luckily, there is a special offer on reformatting json today ;-)

Phemus commented on unit_tests/test_app.py:
```diff
@@ -77,3 +78,7 @@ def test_get_endpoint_not_found(self, mock_query):
         self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_schema_checks(self):
+
```
should this space be here? I don't think one new line breaks pep8 per se but other functions we have implemented always start directly under the definition

Kirosoft replied:
this is o.k. because it is a def within a class - hence one line only - according to flake8

Phemus commented on unit_tests/schema_tests.py:
```diff
@@ -0,0 +1,78 @@
+
+
+def run_schema_checks():
+
```
same as other comment regarding the new line

Kirosoft committed:
Added new tests and removed duplicates based on PR comments
Updated swagger/schema to allow spaces in surnames
Updated schema tests to allow spaces in surnames
Reformatted json for correct spacing
Flake8

Kirosoft committed:
Removed schema.json file. It is now dynamically generated from the swagger json

Kirosoft committed:
Fixed some references to validate swagger via the online editor (added error boject, ensure md_ref case was consistent)

Kirosoft committed:
Update generated schema to use required props from schema definition

tsangsimon commented:
:+1:  good stuff
